### DISCUSSION TITLE
Continuously read/update UserDMA status

### DISCUSSION THREAD

**Discussion by vic9112** [bol-edu/HLS-SOC-Discussions#224] [Q&A] [answered]
Hello everyone, I am currently working on the DMA part of my final project. I want my DMA buffer status to be continuously updated. The first image below is my co-simulation waveform. `icmp` is the `buf_sts`(buffer status) inside the DMA. We can see that after the s2m(Stream-to-AXIMaster) stream transfer is completed, it is pulled to `1`. However, since my interface used block-level protocol, my `s2m_buf_sts` actually only updates when `ap_done` (also, `s2m_buf_sts_vld`). Therefore, from the interface perspective, after `s2m_buf_sts` changes to `1` for the first time, although the second `ap_start` has begun, it remains at `1`.
![waveform](https://github.com/bol-edu/HLS-SOC-Discussions/assets/137171415/ecdeb332-1561-4b1b-bf89-61f7ec38ada5)

Top function:
![interface](https://github.com/bol-edu/HLS-SOC-Discussions/assets/137171415/be3daa3f-b234-4f85-8564-124ebe285c97)

`buf_sts` changes in do-while loop:
![buf-sts=1](https://github.com/bol-edu/HLS-SOC-Discussions/assets/137171415/100f7379-0cba-4dff-bfd5-7211194921b2)

**Comment by vic9112** [comment, marked as answer]
I have solved this problem, following is the current waveform:

![waveform](https://github.com/bol-edu/HLS-SOC-Discussions/assets/137171415/6fe2688b-4bb2-4f6d-bb2f-c85b94c60103)

## Interface

![interface](https://github.com/bol-edu/HLS-SOC-Discussions/assets/137171415/ea43a2c6-56b8-4d5e-a48b-94d1085c15ec)

The `volatile` keyword is used to inform the compiler that a variable's value may change at **any time**.

## Reason for using `volatile`
1. Prevent Optimization: 
    Compilers often optimize code by caching the values of variables in registers, assuming they do not change unless explicitly modified by the program. When dealing with hardware signals or memory-mapped I/O, the value can change outside the program's control (e.g., due to hardware events). Using volatile prevents the compiler from making such optimizations, ensuring that the code reads the variable from memory every time it is accessed.
2. Correct Signal Monitoring: 
   In HLS, when you are continuously checking a signal, such as waiting for a hardware event to occur, the signal might be updated by external hardware. Declaring the signal as volatile ensures that the compiler generates code to repeatedly read the actual signal value from the hardware, rather than using a potentially outdated value cached in a register.
3. Synchronization: 
   Hardware often operates asynchronously with respect to the software running on a processor. volatile ensures that the software correctly synchronizes with the hardware, reading the most up-to-date value of a signal, which is crucial for accurate and reliable interaction with hardware components.

**Comment by boledulab** [comment]
Great findings. This is the case with multi-access pointers. The argument s2m_buf_sts is accessed/updated multiple times. The synthesis tool is optimized away unless it is declared volatile. Refer to the PowerPoint below.

[Multi-Access Pointers and RTL Simulation.pptx](https://github.com/bol-edu/HLS-SOC-Discussions/files/15492352/Multi-Access.Pointers.and.RTL.Simulation.pptx)
